azurerm/resource_arm_kubernetes_cluster.go

```diff
@@ -715,6 +724,117 @@ func resourceArmKubernetesClusterDelete(d *schema.ResourceData, meta interface{}
 	return nil
 }
 
+func azureMonitorRoleAssignment(d *schema.ResourceData, meta interface{}) {
+	servicePrincipalProfile := expandAzureRmKubernetesClusterServicePrincipal(d)
+
+	// Get the root roleDefinitionID full path (i.e. providers/Microsoft.Authorization/roleDefinitions/3913510d-42f4-4e42-8a64-420c390055eb)
+	rootRoleDefinitionID := getRootRoleDefinitionID(meta)
+
+	if rootRoleDefinitionID == "" {
+		log.Printf("[INFO] Unable to locate 'Monitoring Metrics Publisher' Root Role Definition Id\n")
+		return
+	}
+
+	// Set the scope to the AKS cluster (i.e. subscriptions/<subscription id>/resourcegroups/<resource group name>/providers/Microsoft.ContainerService/managedClusters/<resource name>)
+	scope := strings.TrimPrefix(d.Id(), "/")
+	log.Printf("[INFO] Role Assignment Scope is: %s\n", scope)
+
+	// UUID Name of the Role Assignment in this scope
+	name := "3913510d-42f4-4e42-8a64-420c390055eb"
+
+	// Build the roleDefinitionID for this subscription (i.e. subscriptions/<subscription id>/providers/Microsoft.Authorization/roleDefinitions/<Role Assignment Name>)
+	roleDefinitionID := fmt.Sprintf("subscriptions/%s%s", *servicePrincipalProfile.ClientID, rootRoleDefinitionID)
+	log.Printf("[INFO] Role Definition ID: %s\n", roleDefinitionID)
+
+	clientServicePrincipalObjectId := getServicePrincipalObjectID(*servicePrincipalProfile.ClientID, meta)
+	log.Printf("[INFO] Service Principal Object ID: %s\n", clientServicePrincipalObjectId)
+
+	if clientServicePrincipalObjectId == "" {
+		log.Printf("[DEBUG] Unable to lookup Service Principal Object ID\n")
+		return
+	}
+
+	properties := authorization.RoleAssignmentCreateParameters{
+		RoleAssignmentProperties: &authorization.RoleAssignmentProperties{
+			// Monitoring Metrics Publisher Role
+			RoleDefinitionID: utils.String(roleDefinitionID),
+			PrincipalID:      utils.String(clientServicePrincipalObjectId),
+		},
+	}
+
+	if err := resource.Retry(300*time.Second, retryAzureMonitorRoleAssignment(scope, name, properties, meta)); err != nil {
+		log.Printf("[DEBUG] Error adding Role Assignment to AKS cluster: %+v\n", err)
```

Comment:
Shouldn't we error out and fail the deployment?

Reply:
> Shouldn't we error out and fail the deployment?

Role assignment requires 'Microsoft.Authorization/roleAssignments/write' permission and this permission only granted to owner and administrator built-in roles by default.  From monitoring side, we want this role assignment optional hence we shouldn't fail the deployment. Without this role assignment, they get other monitoring data.

Reply:
However silently failing does not provide any feedback to the user expecting the role to be set. it becomes non obvious as to why metrics are not being published.

I think a new property `assign_monitor_role` separating out the option so that if it does fail the user gets an error messaging suggestion actions they can take an if its not something they want it cant just be unset it and continue with their deployment.